File: code/client/munkilib/reports.py
# encoding: utf-8
#
# Copyright 2009-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
reports.py

Created by Greg Neagle on 2016-12-14.


Reporting functions
"""
from __future__ import absolute_import, print_function

import os
import subprocess
import sys
import time

# PyLint cannot properly find names inside Cocoa libraries, so issues bogus
# No name 'Foo' in module 'Bar' warnings. Disable them.
# pylint: disable=E0611
from Foundation import NSDate
# pylint: enable=E0611

from . import munkilog
from . import prefs
from . import FoundationPlist


def format_time(timestamp=None):
    """Return timestamp as an ISO 8601 formatted string, in the current
    timezone.
    If timestamp isn't given the current time is used."""
    if timestamp is None:
        return str(NSDate.new())
    return str(NSDate.dateWithTimeIntervalSince1970_(timestamp))


def printreportitem(label, value, indent=0):
    """Prints a report item in an 'attractive' way"""
    indentspace = '    '
    if isinstance(value, type(None)):
        print(indentspace*indent, '%s: !NONE!' % label)
    elif isinstance(value, list) or type(value).__name__ == 'NSCFArray':
        if label:
            print(indentspace*indent, '%s:' % label)
        index = 0
        for item in value:
            index += 1
            printreportitem(index, item, indent+1)
    elif isinstance(value, dict) or type(value).__name__ == 'NSCFDictionary':
        if label:
            print(indentspace*indent, '%s:' % label)
        for subkey in value.keys():
            printreportitem(subkey, value[subkey], indent+1)
    else:
        print(indentspace*indent, '%s: %s' % (label, value))


def printreport(reportdict):
    """Prints the report dictionary in a pretty(?) way"""
    for key in reportdict.keys():
        printreportitem(key, reportdict[key])


def savereport():
    """Save our report"""
    FoundationPlist.writePlist(
        report, os.path.join(
            prefs.pref('ManagedInstallDir'), 'ManagedInstallReport.plist'))


def readreport():
    """Read report data from file"""
    global report
    reportfile = os.path.join(
        prefs.pref('ManagedInstallDir'), 'ManagedInstallReport.plist')
    try:
        report = FoundationPlist.readPlist(reportfile)
    except FoundationPlist.NSPropertyListSerializationException:
        report = {}


def _warn(msg):
    """We can't use display module functions here because that would require
    circular imports. So a partial reimplementation."""
    warning = 'WARNING: %s' % msg
    print(warning.encode('UTF-8'), file=sys.stderr)
    munkilog.log(warning)
    # append this warning to our warnings log
    munkilog.log(warning, 'warnings.log')


def archive_report():
    """Archive a report"""
    reportfile = os.path.join(
        prefs.pref('ManagedInstallDir'), 'ManagedInstallReport.plist')
    if os.path.exists(reportfile):
        modtime = os.stat(reportfile).st_mtime
        formatstr = '%Y-%m-%d-%H%M%S'
        archivename = ('ManagedInstallReport-%s.plist'
                       % time.strftime(formatstr, time.localtime(modtime)))
        archivepath = os.path.join(prefs.pref('ManagedInstallDir'), 'Archives')
        if not os.path.exists(archivepath):
            try:
                os.mkdir(archivepath)
            except (OSError, IOError):
                _warn('Could not create report archive path.')
        try:
            os.rename(reportfile, os.path.join(archivepath, archivename))
        except (OSError, IOError):
            _warn('Could not archive report.')
        # now keep number of archived reports to 100 or fewer
        proc = subprocess.Popen(['/bin/ls', '-t1', archivepath],
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE)
        output = proc.communicate()[0].decode('UTF-8')
        if output:
            archiveitems = [item
                            for item in str(output).splitlines()
                            if item.startswith('ManagedInstallReport-')]
            if len(archiveitems) > 100:
                for item in archiveitems[100:]:
                    itempath = os.path.join(archivepath, item)
                    if os.path.isfile(itempath):
                        try:
                            os.unlink(itempath)
                        except (OSError, IOError):
                            _warn('Could not remove archive item %s' % item)


# module globals
# pylint: disable=invalid-name
report = {}
# pylint: enable=invalid-name


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/reports.swift
//
//  reports.swift
//  munki
//
//  Created by Greg Neagle on 7/1/24.
//
//  Copyright 2024-2026 The Munki Project. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation

/// A Singleton class to manage reports
class Report {
    static let shared = Report()

    var report: PlistDict

    private init() {
        report = PlistDict()
    }

    func record(_ value: Any, to key: String) {
        report[key] = value
    }

    func retrieve(key: String) -> Any? {
        return report[key]
    }

    func add(string newValue: String, to key: String) {
        if var value = report[key] as? [String] {
            value.append(newValue)
            report[key] = value
        } else {
            report[key] = [newValue]
        }
    }

    func add(dict newValue: PlistDict, to key: String) {
        if var value = report[key] as? [PlistDict] {
            value.append(newValue)
            report[key] = value
        } else {
            report[key] = [newValue]
        }
    }

    /// Returns path to report file
    func reportFile() -> String {
        return managedInstallsDir(subpath: "ManagedInstallReport.plist")
    }

    /// Saves our report
    func save() {
        do {
            try writePlist(report, toFile: reportFile())
        } catch {
            DisplayAndLog.main.error(
                "Failed to write ManagedInstallReport.plist: \(error)"
            )
        }
    }

    /// Read report data from file
    func read() {
        do {
            if let temp = try readPlist(fromFile: reportFile()) {
                report = temp as? PlistDict ?? PlistDict()
            }
        } catch {
            report = PlistDict()
        }
    }

    /// Archive current report file
    func archiveReport() {
        let display = DisplayAndLog.main
        let reportFile = managedInstallsDir(subpath: "ManagedInstallReport.plist")
        if !pathExists(reportFile) {
            // nothing to do
            return
        }
        let dateformatter = DateFormatter()
        dateformatter.dateFormat = "yyyy-MM-dd-HHmmss"
        let timestamp = dateformatter.string(from: Date())
        let archiveName = "ManagedInstallReport-\(timestamp).plist"
        let archiveDir = managedInstallsDir(subpath: "Archives")
        if !pathExists(archiveDir) {
            do {
                try FileManager.default.createDirectory(atPath: archiveDir, withIntermediateDirectories: false)
            } catch {
                display.warning("Could not create report archive directory: \(error.localizedDescription)")
            }
        }
        let fullArchivePath = (archiveDir as NSString).appendingPathComponent(archiveName)
        do {
            try FileManager.default.moveItem(atPath: reportFile, toPath: fullArchivePath)
        } catch {
            display.warning("Could not archive report: \(error.localizedDescription)")
        }
        // now keep number of archived reports to 100 or fewer
        let directoryURL = URL(fileURLWithPath: archiveDir)
        do {
            let contents = try
                FileManager.default.contentsOfDirectory(
                    at: directoryURL,
                    includingPropertiesForKeys: [.contentModificationDateKey],
                    options: [.skipsHiddenFiles, .skipsSubdirectoryDescendants]
                )
                .filter { $0.lastPathComponent.hasPrefix("ManagedInstallReport-") }
                .sorted(by: {
                    let date0 = try $0.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate!
                    let date1 = try $1.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate!
                    return date0.compare(date1) == .orderedDescending
                })
            if contents.count > 100 {
                for item in contents[100...] {
                    do {
                        try FileManager.default.removeItem(at: item)
                    } catch {
                        display.warning("Error removing \(item.path): \(error.localizedDescription)")
                    }
                }
            }
        } catch {
            display.warning("Error accessing archived report directory: \(error.localizedDescription)")
        }
    }
}
